README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rp5pik <img src="man/figures/rp5pik_logo.png" align="right" height="139" />

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![pkgcheck](https://github.com/atsyplenkov/rp5pik/workflows/pkgcheck/badge.svg)](https://github.com/atsyplenkov/rp5pik/actions?query=workflow%3Apkgcheck)
[![CRAN
status](https://www.r-pkg.org/badges/version/rp5pik)](https://CRAN.R-project.org/package=rp5pik)
![GitHub R package
version](https://img.shields.io/github/r-package/v/atsyplenkov/rp5pik?label=github)
![GitHub last
commit](https://img.shields.io/github/last-commit/atsyplenkov/rp5pik)
<!-- badges: end -->

The `rp5pik` package provides a set of functions to download and
preprocess meteorological data from <http://www.pogodaiklimat.ru/>

## Installation

You can install the development version of rp5pik from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("atsyplenkov/rp5pik")

# OR

# install.packages("remotes")
remotes::install_github("atsyplenkov/rp5pik")
```

## Examples

### 1. Data download

Below is an example for `rp_parse_pik` functions. It allows you to
download meteo data at **3-hour** temporal resolution for various
stations using their WMO ID from <http://www.pogodaiklimat.ru/>:

``` r
library(rp5pik)

example <-
  rp_parse_pik(
    wmo_id = c("20069", "27524"),
    start_date = "2022-05-01",
    end_date = "2022-05-31"
  )
#> Parsing data

example
#> # A tibble: 496 × 11
#>    wmo   datetime_utc           ta    td    rh    ps   psl  prec windd winds_m…¹
#>    <chr> <dttm>              <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <int>     <dbl>
#>  1 20069 2022-05-01 00:00:00  -7.9  -9.1    91  993   996.   0.8    90         6
#>  2 20069 2022-05-01 03:00:00  -7    -8.2    91  994.  997.  NA      90         7
#>  3 20069 2022-05-01 06:00:00  -7.2  -8.4    91  994.  997.  NA      90         9
#>  4 20069 2022-05-01 09:00:00  -7.7  -8.8    92  994.  998.  NA      90        11
#>  5 20069 2022-05-01 12:00:00  -8.4  -9.5    92  995.  998.   0.6    90        11
#>  6 20069 2022-05-01 15:00:00  -9.8 -11      91  995   998.  NA      45        11
#>  7 20069 2022-05-01 18:00:00 -11.3 -12.7    89  995.  998.  NA      45         9
#>  8 20069 2022-05-01 21:00:00 -14.2 -16      86  995.  999.  NA      45         7
#>  9 20069 2022-05-02 00:00:00 -14.3 -15.9    88  995.  998.  NA      45         6
#> 10 20069 2022-05-02 03:00:00 -11.8 -14.4    81  994.  997.  NA       0         6
#> # … with 486 more rows, 1 more variable: winds_max <dbl>, and abbreviated
#> #   variable name ¹​winds_mean
```

List of available variables:

- `wmo` character. WMO index of the meteostation

- `datetime_utc` POSIXct. Date and Time of the measurement at UTC

- `ta` numeric. Air temperature at 2m above the surface, °C

- `td` numeric. Dew point, °C

- `rh` numeric. Relative humidity at 2m above the surface, %

- `ps` numeric. Atmosphere pressure at meteostation, hPa

- `psl` numeric. Atmosphere pressure adjusted to the height of mean sea
  level, hPa

- `prec` numeric. Cumulative precipitation for the last 12 hours, mm

- `windd` integer. Wind direction, deg

- `winds_mean` numeric. Average 10-min wind speed, m/s

- `winds_max` numeric. Maximum wind speed, m/s

We can visualize the `example` dataset using `ggplot2` as follows:

``` r
library(ggplot2)

example |> 
  ggplot(
    aes(
      x = datetime_utc,
      y = ta,
      group = wmo
    )
  ) +
  geom_line(aes(color = wmo)) +
  labs(
    x = "",
    y = "Average Temperature, °C"
  ) +
  theme_minimal()
```

<img src="man/figures/README-plot-1.png" width="80%" style="display: block; margin: auto;" />

### 2. Data preprocessing

Since the downloaded with `rp_parse_pik` data contains raw data, it
requires additional checking and cleaning. We suggest to explore the raw
dataset by yourselves before any further manipulations.

However, the `rp5pik` package has a function to aggregate raw data on
daily (`24h`) or semi-daily (`12h`) periods. The `rp_aggregate_pik`
function removes known error codes from precipitation data (`699`
values). Additionally, it calculates daily precipitation sums based on
measured precipitation at 06 UTC and 18 UTC in European part of Russia
(see [meteostation manuals for more
info](https://method.meteorf.ru/ansambl/pojasnenijaansambl.html)).

⚠ As of 2023-03-17 this function works only with Moscow timezone.

This is how you can aggregate data daily:

``` r
library(dplyr)

example_daily <- 
  example |> 
  rp_aggregate_pik(.period = "24h") |> 
  group_split(wmo)

example_daily
#> <list_of<
#>   tbl_df<
#>     wmo       : character
#>     date      : date
#>     p         : double
#>     ta        : double
#>     td        : double
#>     rh        : double
#>     ps        : double
#>     psl       : double
#>     windd     : double
#>     winds_mean: double
#>     winds_max : double
#>   >
#> >[2]>
#> [[1]]
#> # A tibble: 32 × 11
#>    wmo   date           p     ta     td    rh    ps   psl windd winds_…¹ winds…²
#>    <chr> <date>     <dbl>  <dbl>  <dbl> <dbl> <dbl> <dbl> <dbl>    <dbl>   <dbl>
#>  1 20069 2022-05-01    NA  -8.47  -9.67  91    994.  998.  77.1     9.14      13
#>  2 20069 2022-05-02    NA -12.9  -14.9   85    992.  995. 129.      6.12      10
#>  3 20069 2022-05-03    NA -16.0  -19.0   77.8  988.  992. 152.      2.25      NA
#>  4 20069 2022-05-04    NA -14.3  -17.4   77.2  996.  999. 248.      5.38      10
#>  5 20069 2022-05-05    NA -14.2  -16.0   86.6  998. 1002.  67.5     3.38      NA
#>  6 20069 2022-05-06    NA -12.6  -14.3   87.2  994.  997. 253.      4.12      NA
#>  7 20069 2022-05-07    NA  -9.31 -11.1   86.9  999. 1002. 231.      3.12      NA
#>  8 20069 2022-05-08    NA  -6.88  -8.25  90.1  999. 1003. 152.      9.25      18
#>  9 20069 2022-05-09    NA  -7.19  -9.02  86.6  991.  995. 202.      9.62      17
#> 10 20069 2022-05-10    NA  -7.25  -8.89  88    987.  990. 197.      7.75      12
#> # … with 22 more rows, and abbreviated variable names ¹​winds_mean, ²​winds_max
#> 
#> [[2]]
#> # A tibble: 32 × 11
#>    wmo   date           p    ta     td    rh    ps   psl windd winds_m…¹ winds…²
#>    <chr> <date>     <dbl> <dbl>  <dbl> <dbl> <dbl> <dbl> <dbl>     <dbl>   <dbl>
#>  1 27524 2022-05-01  NA    7.21 -5.59   43    996. 1023.  238.      3.71      NA
#>  2 27524 2022-05-02   0    9.85 -2.4    44.1  991. 1018.  199.      3         NA
#>  3 27524 2022-05-03  NA   11.6   1.54   53.9  982. 1008.  231.      5.88      13
#>  4 27524 2022-05-04   0.6  6    -1.14   60.5  985. 1012.  298.      5.25      14
#>  5 27524 2022-05-05  NA    5.24 -6.68   45.5  997. 1025.  248.      3.75      NA
#>  6 27524 2022-05-06  NA    9.59 -5.91   34.2 1001. 1028.  191.      3.5       NA
#>  7 27524 2022-05-07  NA   13.1  -2.59   34.2  999. 1025.  180       3.62      11
#>  8 27524 2022-05-08   1   13.6   2.17   47.8  992. 1019.  208.      4.75      11
#>  9 27524 2022-05-09   2.6  6.99  0.575  64.2  994. 1021.  158.      6.25      14
#> 10 27524 2022-05-10   0    6.62 -3.5    50.6  992. 1019.  158.      7.38      17
#> # … with 22 more rows, and abbreviated variable names ¹​winds_mean, ²​winds_max
```

Or semi-daily:

``` r
library(dplyr)

example_12h <- 
  example |> 
  rp_aggregate_pik(.period = "12h", .tz = "Europe/Moscow") |> 
  group_split(wmo)

example_12h
#> <list_of<
#>   tbl_df<
#>     wmo        : character
#>     datetime_tz: datetime<Europe/Moscow>
#>     p          : double
#>     ta         : double
#>     td         : double
#>     rh         : double
#>     ps         : double
#>     psl        : double
#>     windd      : double
#>     winds_mean : double
#>     winds_max  : double
#>   >
#> >[2]>
#> [[1]]
#> # A tibble: 63 × 11
#>    wmo   datetime_tz             p     ta     td    rh    ps   psl windd winds…¹
#>    <chr> <dttm>              <dbl>  <dbl>  <dbl> <dbl> <dbl> <dbl> <dbl>   <dbl>
#>  1 20069 2022-05-01 09:00:00    NA  -7.37  -8.57  91    993.  997.  90      7.33
#>  2 20069 2022-05-01 21:00:00    NA  -9.3  -10.5   91    995.  998.  67.5   10.5 
#>  3 20069 2022-05-02 09:00:00    NA -13.0  -15.0   84.8  994.  997.  22.5    6.25
#>  4 20069 2022-05-02 21:00:00    NA -12.8  -14.8   85.2  989.  993. 236.     6   
#>  5 20069 2022-05-03 09:00:00    NA -17.0  -19.7   79.8  987.  990.  90      2   
#>  6 20069 2022-05-03 21:00:00    NA -15    -18.4   75.8  990.  993. 214.     2.5 
#>  7 20069 2022-05-04 09:00:00    NA -15.0  -18.6   74.5  993.  997. 248.     5.75
#>  8 20069 2022-05-04 21:00:00    NA -13.5  -16.2   80    998. 1001. 248.     5   
#>  9 20069 2022-05-05 09:00:00    NA -14.9  -16.6   86.8 1000. 1003. 124.     1.5 
#> 10 20069 2022-05-05 21:00:00    NA -13.4  -15.3   86.5  997. 1000.  11.2    5.25
#> # … with 53 more rows, 1 more variable: winds_max <dbl>, and abbreviated
#> #   variable name ¹​winds_mean
#> 
#> [[2]]
#> # A tibble: 63 × 11
#>    wmo   datetime_tz             p    ta    td    rh    ps   psl windd winds_m…¹
#>    <chr> <dttm>              <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>     <dbl>
#>  1 27524 2022-05-01 09:00:00  NA    2.47 -5.27  59.3  998. 1026.  225       2.33
#>  2 27524 2022-05-01 21:00:00  NA   10.8  -5.82  30.8  995. 1022.  248.      4.75
#>  3 27524 2022-05-02 09:00:00  NA    7.55 -1.58  53    993. 1020.  195       1.5 
#>  4 27524 2022-05-02 21:00:00   0   12.2  -3.22  35.2  990. 1016.  202.      4.5 
#>  5 27524 2022-05-03 09:00:00  NA    7.8   3.22  73.2  985. 1011.  214.      3.25
#>  6 27524 2022-05-03 21:00:00  NA   15.4  -0.15  34.5  980. 1006.  248.      8.5 
#>  7 27524 2022-05-04 09:00:00   0.3  7     0.55  63.5  982. 1009.  292.      4   
#>  8 27524 2022-05-04 21:00:00   0.3  5    -2.82  57.5  988. 1015.  304.      6.5 
#>  9 27524 2022-05-05 09:00:00  NA    1.9  -4.85  62    996. 1023.  315       3.75
#> 10 27524 2022-05-05 21:00:00  NA    8.58 -8.5   29    999. 1026.  180       3.75
#> # … with 53 more rows, 1 more variable: winds_max <dbl>, and abbreviated
#> #   variable name ¹​winds_mean
```

## Roadmap

    rp5pik 📦
    ├── Parser functions for
    │   ├── pogodaiklimat
    │   │   ├── rp5pik::rp_parse_pik ✅
    │   │   └── rp5pik::rp_aggregate_pik ✅
    │   ├── rp5 🔲
    │   └── gmvo.skniivh 🔲
    ├── WMO stations coordinates  🔲
    └── Rain/Snow guessing  
        └── rp5pik::rp_get_temp50 ✅


NEWS.md
# rp5pik 0.0.2.2

* The previous update contained a critical issue with time aggregating, therefore the `.period` param from `rp_aggregate_pik` has been removed. Currently the aggregation in semi-daily format is working only in left direction, i.e. it calculates a mean value of all params before the timestamp.

# rp5pik 0.0.2.1

* Added new variable to `rp_aggregate_pik` controlling the aggregation direction: left or center. Still works only with "Europe/Moscow" timezone

# rp5pik 0.0.2

* Minor improvements in `rp_parse_pik` description and dependencies.
* Added function `rp_aggregate_pik` to summarise meteo data on a daily or 12-hours basis
* New function `rp_get_temp50` for detecting a 50% rain-snow air temperature treshold based on the work by *Jennings et al. ([2018](https://www.nature.com/articles/s41467-018-03629-7))*


# rp5pik 0.0.1

* Initial version with `rp_parse_pik` functionality.


R/parse_pik.R
#' Parse pogodaiklimat.ru
#'
#' `r lifecycle::badge('stable')`
#'
#' @description A function to parse [http://www.pogodaiklimat.ru/](http://www.pogodaiklimat.ru/)
#' 3-hourly meteo data.
#'
#' @param wmo_id character. WMO meteostation ID to parse
#' @param start_date character. The beginning of the parsing
#' period. Write date in the following format `YYYY-MM-DD`
#' @param end_date character. The end of the parsing period
#' period. Write date in the following format `YYYY-MM-DD`
#'
#' @details
#' The function returns a [tibble::tibble()] with
#' following variables:
#'
#'  - `wmo` -- character. WMO index of the meteostation 3&deg;C
#'  - `datetime_utc` -- POSIXct. Date and Time of the measurement at UTC
#'  - `ta` -- numeric. Air temperature at 2m above the surface, &deg;C
#'  - `td` -- numeric. Dew point, &deg;C
#'  - `rh` -- numeric. Relative humidity at 2m above the surface, %
#'  - `ps` -- numeric. Atmosphere pressure at meteostation, hPa
#'  - `psl` -- numeric. Atmosphere pressure adjusted to the height of mean sea level, hPa
#'  - `prec` -- numeric. **Cumulative** precipitation for the last **12 hours**, mm
#'  - `windd` -- integer. Wind direction, &deg;C
#'  - `winds_mean` -- numeric. Average 10-min wind speed, m/s
#'  - `winds_max` -- numeric. Maximum wind speed, m/s
#'
#' @return A [tibble::tibble()]
#' @export
#'
#' @examples
#' library(rp5pik)
#'
#' example <-
#'   parse_pik(
#'     wmo_id = c("20069", "27524"),
#'     start_date = "2022-05-01",
#'     end_date = "2022-05-31"
#'   )
#'
#' example
#'
#' @import cli rvest
#' @importFrom stats time
#' @importFrom lubridate day days_in_month month year make_datetime with_tz force_tz
#' @importFrom stringr str_split
#' @importFrom purrr map_df
#' @importFrom tidyr unnest
#' @importFrom dplyr mutate select
#'

parse_pik <-
  function(wmo_id = "27524",
           start_date = "2022-04-01",
           end_date = "2022-05-31"){

    # Create a sequence of monthes
    .dates <-
      base::seq.Date(
        from = base::as.Date(start_date),
        to = base::as.Date(end_date),
        by = 'month'
      )

    # Helper variables
    # Empty dataframe to write data to
    .all_df <- base::data.frame()
    # Amount of iterations
    .tot_blocks <- base::length(wmo_id) * base::length(.dates)

    # Initialize the progress bar
    cli::cli_progress_bar(
      "Downloading data",
      type = "tasks",
      total = .tot_blocks
    )

    # Double for-loop
    for(i in wmo_id){
      for(d in .dates){

        # create url to parse
        d <-
          base::as.Date(d,
                  origin = '1970-01-01')
        u <-
          base::paste0(
            'http://www.pogodaiklimat.ru/weather.php?id=', i,
            '&bday=', lubridate::day(d),
            '&fday=', lubridate::days_in_month(d),
            '&amonth=', lubridate::month(d),
            '&ayear=', lubridate::year(d),
            '&bot=2'
          )

        pik_one <-
          tryCatch(
            {
              .parse_pik(u, i, lubridate::year(d))
            },
            error = function(e) {
              cli::cli_inform(
                c("i" = "No data at station {i} for {lubridate::month(d, label = TRUE, abbr = FALSE, locale = 'en_us')} {lubridate::year(d)}")
              )
              return(NULL)
            },
            warning = function(w) {
              return(NULL)
            }
          )

        # Add downloaded data to an empty dataframe
        .all_df <- base::rbind(.all_df, pik_one)

        # Progress bar update
        cli::cli_progress_update()
      }
    }

    # End of progress bar
    cli::cli_progress_done()

    if (nrow(.all_df) == 0) {
      cli::cli_abort(c(
        "Your {.var start_date} or {.var end_date} are out of the data range available on the http://www.pogodaiklimat.ru/",
        "x" = "There is no data available from {start_date} to {end_date} period",
        "i" = "Try different period or reduce the amount of stations"
      ))
    }

    .all_df |>
      dplyr::mutate(wind = purrr::map_df(winds,
                                         ~.parse_winds(.x))) |>
      tidyr::unnest(cols = c(wind)) |>
      dplyr::mutate(windd = .parse_windd(windd)) |>
      dplyr::mutate(datetime_utc =
                      lubridate::make_datetime(
                        year = year,
                        month = base::as.integer(stringr::str_split(daymon, "\\.", simplify = T)[,2]),
                        day = base::as.integer(stringr::str_split(daymon, "\\.", simplify = T)[,1]),
                        hour = time, tz = "UTC"
                      ),
                    .before = "ta"
      )  |>
      dplyr::select(
        wmo,
        datetime_utc,
        ta,
        td,
        rh,
        ps,
        psl,
        prec,
        windd,
        winds_mean,
        winds_max
      )

  }

#' Initial parsing function
#'
#' @import rvest
#'
#' @noRd

.parse_pik <-
  function(url, ind, yr){

    tables <-
      rvest::html_nodes(
        rvest::read_html(url), "table"
      )  # парсим по тэгу

    tbl1 <-
      rvest::html_table(tables[[1]],
                        header = F) # список из двух таблиц, в одной - сроки и дата, во второй - данные ??\_(???)_/??
    tbl2 <-
      rvest::html_table(tables[[2]],
                        header = F)

    tbl1 <- tbl1[-1,] # избавляемся от ненужных хэдеров
    tbl2 <- tbl2[-1,]

    # записываем в фрейм
    df <-
      base::data.frame(
        wmo = ind,
        year = yr,
        daymon = base::as.character(tbl1$X2),
        time = base::as.numeric(tbl1$X1),
        ta = base::as.numeric(tbl2$X6),
        td = base::as.numeric(tbl2$X7),
        rh = base::as.numeric(tbl2$X8),
        ps = base::as.numeric(tbl2$X13),
        psl = base::as.numeric(tbl2$X12),
        winds = tbl2$X2,
        windd = tbl2$X1,
        prec = base::as.numeric(tbl2$X16)
      )

    return(df)

  }

#' Parse wind speed
#'
#' @importFrom stringr str_detect
#' @importFrom tibble tibble
#'
#' @noRd

.parse_winds <-
  function(x){

    # Conditions
    con_storm <-
      stringr::str_detect(x, "\\{") & !stringr::str_detect(x, "(-)")
    con_mean <-
      stringr::str_detect(x, "(-)") & !stringr::str_detect(x, "\\{")
    con_mean_storm <-
      stringr::str_detect(x, "(-)") & stringr::str_detect(x, "\\{")

    if (con_storm) {

      .values <-
        .extract_values(x)

      .df <-
        tibble::tibble(
          winds_mean = .values[1],
          winds_max = NA_real_,
          winds_nesrok = .values[2]
        )

      return(.df)

    } else if (con_mean) {

      .values <-
        .extract_values(x)

      .df <-
        tibble::tibble(
          winds_mean = .values[1],
          winds_max = .values[2],
          winds_nesrok = NA_real_
        )

      return(.df)

    } else if (con_mean_storm) {

      .values <-
        .extract_values(x)

      .df <-
        tibble::tibble(
          winds_mean = .values[1],
          winds_max = .values[2],
          winds_nesrok = .values[3]
        )

      return(.df)

    } else {

      .df <-
        tibble::tibble(
          winds_mean = base::as.numeric(x),
          winds_max = NA_real_,
          winds_nesrok = NA_real_
        )

      return(.df)

    }

  }

#' Extract values from a string with wind speed
#'
#' @noRd

.extract_values <-
  function(string) {

    values <- base::strsplit(string, "[{}|-]")[[1]]
    return(base::as.numeric(values))

  }

#' Parse wind direction
#'
#' @importFrom tibble tibble
#'
#' @noRd

.parse_windd <-
  function(x) {

    # С - северный,
    # СВ - северо-восточный,
    # В - восточный,
    # ЮВ - юго-восточный,
    # Ю - южный,
    # ЮЗ - юго-западный,
    # З - западный,
    # СЗ - северо-западный
    # нст - нет данных (?)

    .trans_table <-
      tibble::tibble(
        dir_char =
          c("С", "СВ", "В", "ЮВ", "Ю",
            "ЮЗ", "З", "СЗ", "нст"),
        dir_deg =
          c(0L, 45L, 90L, 135L, 180L,
            225L, 270L, 315L, NA_integer_)
      )

    matches <- base::match(x, .trans_table$dir_char)
    result <- .trans_table$dir_deg[matches]

    return(result)

  }



R/get_temp50.R
#' Get 50% rain-snow air temperature threshold
#'
#' `r lifecycle::badge('experimental')`
#'
#' @description This function uses 50% rain-snow air temperature threshold map
#' created by *Jennings et al.* ([2018](https://www.nature.com/articles/s41467-018-03629-7))
#' and returns the expected threshold value in a point. Works
#' only for Northern Hemisphere. See *Jennings et al.* ([2018](https://www.nature.com/articles/s41467-018-03629-7))
#' for coverage details.
#'
#' At this temperature precipitation occurs as rain and snow
#' with equal frequency, while above the threshold precipitation
#' is primarily rain and below primarily snow.
#'
#' @param poi SpatVector. A point layer for the Point Of Interest
#'
#' @return A data.frame
#'
#' @references Jennings, K.S., Winchell, T.S., Livneh, B. et al. Spatial variation of the rain–snow temperature threshold across the Northern Hemisphere. Nat Commun 9, 1148 (2018). https://doi.org/10.1038/s41467-018-03629-7
#'
#' @export
#'
#' @import cli
#' @importFrom terra vect project extract rast

get_temp50 <-
  function(
    poi
  ) {

    if (base::any(base::class(poi) == c("sf", "SpatialPointsDataFrame"))) {

      aoi <- terra::vect(poi)
      aoi <- terra::project(aoi, "EPSG:4326")

    } else if (base::any(base::class(poi) == "SpatVector")) {

      aoi <- poi
      aoi <- terra::project(aoi, "EPSG:4326")

    }

    rs_rast <-
      terra::rast("inst/extdata/jennings_et_al_2018_temp50_cog.tiff")

    base::names(rs_rast) <- "temp50"

    rs_rast <- rs_rast / 10^4

    terra::extract(rs_rast, aoi)

  }


# todo --------------------------------------------------------------------

# get_temp50 <-
#   function(
    #     wmo_id = "29231",
#     lat,
#     lon,
#     sf_obj
#   ) {
#
#     # load observed 50% rain-snow threshold
#     obs_file <- "inst/extdata/jennings_et_al_2018_file3_temp50_observed_by_station.Rds"
#       # system.file("extdata2/jennings_et_al_2018_file3_temp50_observed_by_station.Rds",
#                   # package = "rp5pik")
#     obs_rs <-
#       base::readRDS(obs_file)
#
#     # Check if wmo_id is provided
#     if (!base::is.null(wmo_id)) {
#
#       obs_rs_wmo <-
#         base::match(
#           base::as.character(wmo_id),
#           obs_rs$Station_ID
#         )
#
#       if (base::is.na(obs_rs_wmo)) {
#
#         cli::cli_abort(
#           c("x" = "There is no measured 50% temperature
#             rain-snow threshold in Jennings et al. (2018).
#             Please, provide {.var sf_obj} or coordinates
#             ({.var lat} and {.var lon})")
#         )
#
#       } else if (!base::is.na(obs_rs_wmo)) {
#
#         return(obs_rs$temp50[obs_rs_wmo])
#
#       } else {
#
#         cli::cli_abort(
#           c("x" = "Unexpected error!")
#         )
#
#       }
#
#     } else if (base::is.null(wmo_id)) {
#
#       cli::cli_abort(
#         c("x" = "Please, provide {.var sf_obj} or coordinates ({.var lat} and {.var lon})")
#       )
#
#     }
#
#   }
#
# get_temp50()




R/rp_get_temp50.R
#' Get 50% rain-snow air temperature threshold
#'
#' `r lifecycle::badge('experimental')`
#'
#' @description This function uses 50% rain-snow air temperature threshold map
#' created by *Jennings et al.*
#' ([2018](https://www.nature.com/articles/s41467-018-03629-7))
#' and returns the expected threshold value in a point. Works
#' only for Northern Hemisphere. See *Jennings et al.*
#' ([2018](https://www.nature.com/articles/s41467-018-03629-7))
#' for coverage details.
#'
#' At this temperature precipitation occurs as rain and snow
#' with equal frequency, while above the threshold precipitation
#' is primarily rain and below primarily snow.
#'
#' @param poi SpatVector. A point layer for the Point Of Interest
#'
#' @return A data.frame
#'
#' @references Jennings, K.S., Winchell, T.S., Livneh, B. et al. Spatial
#' variation of the rain–snow temperature threshold across the Northern
#' Hemisphere. Nat Commun 9, 1148 (2018).
#' https://doi.org/10.1038/s41467-018-03629-7
#'
#' @export
#'
#' @import cli
#' @importFrom terra vect project extract rast

rp_get_temp50 <-
  function(
    poi
  ) {

    if (base::any(base::class(poi) == c("sf", "SpatialPointsDataFrame"))) {

      aoi <- terra::vect(poi)
      aoi <- terra::project(aoi, "EPSG:4326")

    } else if (base::any(base::class(poi) == "SpatVector")) {

      aoi <- poi
      aoi <- terra::project(aoi, "EPSG:4326")

    }

    rs_file <-
      system.file("extdata", "jennings_et_al_2018_temp50_cog.tiff",
                  package = "rp5pik")

    rs_rast <-
      terra::rast(rs_file)

    base::names(rs_rast) <- "temp50"

    rs_rast <- rs_rast / 10^4

    terra::extract(rs_rast, aoi)

  }


# todo --------------------------------------------------------------------

# get_temp50 <-
#   function(
    #     wmo_id = "29231",
#     lat,
#     lon,
#     sf_obj
#   ) {
#
#     # load observed 50% rain-snow threshold
#     obs_file <- "inst/extdata/jennings_et_al_2018_file3_temp50_observed_by_station.Rds"
#       # system.file("extdata2/jennings_et_al_2018_file3_temp50_observed_by_station.Rds",
#                   # package = "rp5pik")
#     obs_rs <-
#       base::readRDS(obs_file)
#
#     # Check if wmo_id is provided
#     if (!base::is.null(wmo_id)) {
#
#       obs_rs_wmo <-
#         base::match(
#           base::as.character(wmo_id),
#           obs_rs$Station_ID
#         )
#
#       if (base::is.na(obs_rs_wmo)) {
#
#         cli::cli_abort(
#           c("x" = "There is no measured 50% temperature
#             rain-snow threshold in Jennings et al. (2018).
#             Please, provide {.var sf_obj} or coordinates
#             ({.var lat} and {.var lon})")
#         )
#
#       } else if (!base::is.na(obs_rs_wmo)) {
#
#         return(obs_rs$temp50[obs_rs_wmo])
#
#       } else {
#
#         cli::cli_abort(
#           c("x" = "Unexpected error!")
#         )
#
#       }
#
#     } else if (base::is.null(wmo_id)) {
#
#       cli::cli_abort(
#         c("x" = "Please, provide {.var sf_obj} or coordinates ({.var lat} and {.var lon})")
#       )
#
#     }
#
#   }
#
# get_temp50()




R/rp_aggregate_pik.R
#' Aggregate parsed PiK data
#'
#' `r lifecycle::badge('experimental')`
#'
#' @description This functions calculates from raw data parsed earlier with
#' the [rp5pik::rp_parse_pik()] function to a daily averages. It
#' took rainfall precipitation sums `prec` measured at 06 UTC and 18 UTC
#' to calculate a daily precipitation sums `p` (For European part of Russia).
#' The temperature `t` and other parameters calculates as a daily average.
#'
#' @param .data dataframe. Outcome of the [rp5pik::rp_parse_pik()]
#' function.
#' @param .period character. Either `12h` or `24h` --
#' a string specifying the desired aggregation period. If it equal to
#' `12h` than an average of all parameters preceeding the timestamp
#' will be returned
#' @param .tz character. A string describing timezone of the
#' meteostation of interest. See [lubridate::with_tz()]
#' for details.
#'
#' @return A [tibble::tibble()]
#' @export
#'
#' @import cli
#' @importFrom tidyr fill complete drop_na everything
#' @importFrom dplyr mutate first last na_if group_by
#' @importFrom dplyr summarise ungroup case_when left_join arrange
#' @importFrom lubridate with_tz hour as_date day month year make_datetime
#'
#' @md


rp_aggregate_pik <-
  function(
    .data,
    .period = c("24h", "12h"),
    .tz = "Europe/Moscow"
  ){

    if (any(base::is.null(.period), base::length(.period) > 1)) {

      cli::cli_abort(
        c(
          "x" = "Please, indicate the desired aggregation
          period {.var .period}: either {.str 24h} or {.str 12h}"
        )
      )

    }

    .data_tz <-
      .data |>
      dplyr::mutate(
        dt = lubridate::with_tz(datetime_utc, .tz)
      ) |>
      dplyr::group_by(wmo) |>
      dplyr::arrange(dt, .by_group = TRUE) |>
      tidyr::complete(dt = base::seq.POSIXt(
        from = dplyr::first(dt),
        to = dplyr::last(dt),
        by = "3 hour"
      )) |>
      dplyr::arrange(dt, .by_group = TRUE) |>
      tidyr::fill(wmo, .direction = "down") |>
      dplyr::ungroup()

    if (.period == "24h") {

      .data_tz |>
        # check for errors
        dplyr::mutate(prec = dplyr::na_if(prec, 699)) |>
        dplyr::mutate(
          p12 = dplyr::case_when(
            # for Europe/Moscow tz only!!!!
            lubridate::hour(dt) %in% c(9, 21) ~ prec,
            TRUE ~ NA_real_
          )
        ) |>
        dplyr::group_by(wmo, date = lubridate::as_date(dt)) |>
        dplyr::summarise(
          p = .sum_na(p12),
          ta = .mean_na(ta),
          td = .mean_na(td),
          rh = .mean_na(rh),
          ps = .mean_na(ps),
          psl = .mean_na(psl),
          windd = .mean_na(windd),
          winds_mean = .mean_na(winds_mean),
          winds_max = .max_na(winds_max),
          .groups = "drop"
        )

    } else if (.period == "12h") {

      .data_12h <-
        .data_tz |>
        mutate(
          datetime_tz = .floor_9h(.dt = dt, .tz = .tz)
        )


      .prec_df <-
        .data_tz |>
        dplyr::mutate(prec = dplyr::na_if(prec, 699)) |>
        dplyr::mutate(
          p12 = dplyr::case_when(
            # for Europe/Moscow tz only!!!!
            lubridate::hour(dt) %in% c(9, 21) ~ prec,
            TRUE ~ NA_real_
          )
        ) |>
        tidyr::drop_na(p12) |>
        dplyr::select(wmo, datetime_tz = dt, p = p12)

      .data_12h |>
        dplyr::group_by(wmo, datetime_tz) |>
        dplyr::summarise(
          # p = .sum_na(p12),
          ta = .mean_na(ta),
          td = .mean_na(td),
          rh = .mean_na(rh),
          ps = .mean_na(ps),
          psl = .mean_na(psl),
          windd = .mean_na(windd),
          winds_mean = .mean_na(winds_mean),
          winds_max = .max_na(winds_max),
          .groups = "drop"
        ) |>
        dplyr::left_join(.prec_df, by = c("wmo", "datetime_tz")) |>
        dplyr::select(wmo, datetime_tz, p, tidyr::everything())

    } else {

      cli::cli_abort(
        c(
          "x" = "Please, indicate the desired aggregation
          period {.var .period}: either {.str 24h} or {.str 12h}"
        )
      )

    }

  }



#' Helper function to calculate correct sum
#'
#' @noRd
.sum_na <-
  function(.v){

    if (base::all(base::is.na(.v))) {

      return(NA_real_)

    } else {

      base::sum(.v, na.rm = T)

    }

  }

#' Helper function to calculate correct mean
#'
#' @noRd
.mean_na <-
  function(.v){

    if (base::all(base::is.na(.v))) {

      return(NA_real_)

    } else {

      base::mean(.v, na.rm = T)

    }

  }

#' Helper function to calculate correct max
#'
#' @noRd
.max_na <-
  function(.v){

    if (base::all(base::is.na(.v))) {

      return(NA_real_)

    } else {

      base::max(.v, na.rm = T)

    }

  }

#' Helper function to floor datetime object
#'
#' @noRd
.floor_9h <-
  function(.dt, .tz){

    .round <-
      dplyr::if_else(lubridate::hour(.dt) %in% c(10:21),
                     "21:00:00", "09:00:00")

    .round <-
      paste0(lubridate::as_date(.dt), " ", .round)

    .round <-
      lubridate::as_datetime(.round, tz = .tz)

    .round <-
      dplyr::if_else(.round < .dt,
                     .round + lubridate::days(1), .round)

  }
